### ISSUE TITLE
Celebratory comment when concept exercise solution 100% matches the exemplar

### ISSUE THREAD

**Issue by angelikatyborska** [exercism/elixir-analyzer#132]
Could we somehow add an analysis step to every concept exercise that will give a celebratory comment when your solution's AST matches 100% the exemplar solution?

It would be nice if we didn't have to copy-paste all of the exemplars to this repo, but somehow fetch them from the track repo. By git submodules?

Would that require adding some CI magic to periodically update the state of the submodule when the track changes so that the analyzer can be deployed often with the new exemplars?

**Comment by jiegillet** [comment]
For most of the concept exercises, we already have the exemplar solution code in the tests, maybe it's simpler that way. Although I do agree, it would be super cool to fetch them automatically. My CI skills are close to 0 unfortunately, but I'm always looking out for things to learn ;)

**Comment by jiegillet** [comment]
Exciting, let's do it :)

**Comment by jiegillet** [comment]
I got the submodule and and automated syncing GitHub Action to work on my forks of `elixir-analyzer` and `elixir` 🎉 
I followed the workflows accepted in the SO link you provided, so that `elixir-analyzer` automatically syncs the submodule whenever a commit is pushed to the `main` branch of `elixir`. 

I still have some uncertainty around creating the GitHub Action tokens for `exercism` rather than for my own account, but here is what I did:

1. Create submodule in `elixir-analyzer` and commit [this file](https://github.com/jiegillet/elixir-analyzer/blob/jie-submodule/.gitmodules)
2. Create Github Action workflow in `elixir-analyzer` and commit [this file](https://github.com/jiegillet/elixir-analyzer/blob/jie-submodule/.github/workflows/elixir_submodule_sync.yml)
3. [Create a Private Access Token](https://docs.github.com/en/github/authenticating-to-github/keeping-your-account-and-data-secure/creating-a-personal-access-token) for my account, select permission "workflow". (this step probably won't work for exercism repo)
4. Copy token, it will need to be pasted into 2 repos
5. `elixir-analyzer` repo > Settings > Secrets > New repository secret > Name: SUBMODULE_TOKEN value: paste token from step 4. (I don't have access to settings for exercism repo)
6. `elixir-analyzer` repo > Actions > Elixir Submodule Sync > run workflow manually
7. If step 6 works and submodule points to latest commit, rejoice. Otherwise GOTO step `:rand.uniform(6)`.
8. Create Github Action workflow in `elixir` and commit [this file](https://github.com/jiegillet/elixir/blob/jie-submodule/.github/workflows/notify_elixir_analyzer.yml).
9. `elixir` repo > Settings > Secrets > New repository secret > Name: SUBMODULE_TOKEN value: paste token from step 4. (I don't have access to settings for exercism repo)
10. `elixir` repo > Actions > Submodule Notify Elixir Analyzer > run workflow manually
11. If step 10 works and makes `elixir-analyzer` Elixir Submodule Sync run, rejoice. Otherwise GOTO `:rand.uniform(10)`.
12. push a commit to `elixir`
13. If `elixir-analyzer` automatically points to new commit, rejoice. Otherwise GOTO step `:rand.uniform(12)`.

I will submit PRs with the submodule and workflow files, but I have no idea how to handle organization-wide secrets. We should probably ping @iHiD?

**Comment by iHiD** [comment]
In the long term, this should probably be at the website-level rather than the analyzer level. But it's good here for now!

The one problem with doing this at the analyzer-level is that I don't really want the Analyzer to deploy every time the Elixir repo gets updated. Also, solutions are locked to git shas, not to the repo head. So rather than syncing the repo, maybe we push the exemplar file into the files that the analyzer gets? I believe, in fact, this already happens, but we'd have to check. Does that sound like a good solution?

**Comment by jiegillet** [comment]
> In the long term, this should probably be at the website-level rather than the analyzer level. But it's good here for now!

Maybe that can be a use of the representer when it goes live? It would be nice to ignore formatting and variable names. 

Pushing the exemplar file would be helpful. We also suspect that it already happens ([from this file](https://github.com/exercism/website/blob/3d7cf1dca23b9f91827d052bfd6f87416bbf8e07/test/commands/submission/analysis/init_test.rb)), but could not confirm.

There is another use we have for the exemplar/example files, which is checking our analyzer tests. It's OK to use the git submodule for that right? We can update it manually and only when we make a change to the analyzer, so there would be a deployment anyway.

**Comment by iHiD** [comment]
> There is another use we have for the exemplar/example files, which is checking our analyzer tests.

Yeah, I don't mind you using a submodule. The key thing is that I don't really want versions of the analyzer deploying when nothing functional has changed. For lots of reasons (harder to debug, more HDD space used on machines, more build time and environmental cost, etc)

You could always check out the repo during CI rather than having it as a submodule too. But I don't have a preference on that:)

**Comment by jiegillet** [comment]
Concerning the last part of this issue that hasn't been solved:

> It would be nice if we didn't have to copy-paste all of the exemplars to this repo, but somehow fetch them from the track repo. By git submodules?

It is now within our reach since we have the submodule, and I did some tests with this a little while back, but haven't managed to find a satisfying solution yet, because the macros are tricky to handle. 

Also, I came to wonder about that feature. I actually do like copy-pasting the exemplar solution in the tests, because it really helps comparing it to alternate solutions and solutions that break the checks, and copy-pasting from it to build those solutions. I also think that if we didn't copy the solution there, it would feel like something's missing, even with some kind of `import_exemplar()` macro somewhere.

Finally, we can now be sure that the solution that we copy-paste there is identical to the one in the repo because of the common check that checks for identical solutions.

Basically I'm wondering if we should pursue this or close the issue.
